server/plugins/task-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/task/obmysql/structurecompare/OdcStructureComparator.java

```diff
@@ -0,0 +1,572 @@
+            List<DBTableIndex> tgtTabIdx) {
+        List<DBObjectComparisonResult> returnVal = new ArrayList<>();
+        if (srcTabIdx.isEmpty() && tgtTabIdx.isEmpty()) {
+            return returnVal;
+        }
+
+        List<String> srcIdxNames = srcTabIdx.stream().map(DBTableIndex::getName).collect(Collectors.toList());
+        List<String> tarIdxNames = tgtTabIdx.stream().map(DBTableIndex::getName).collect(Collectors.toList());
+        Map<String, DBTableIndex> srcIdxMapping =
+                srcTabIdx.stream().collect(Collectors.toMap(DBTableIndex::getName, col -> col));
+        Map<String, DBTableIndex> tarIdxMapping =
+                tgtTabIdx.stream().collect(Collectors.toMap(DBTableIndex::getName, col -> col));
+
+        tarIdxNames.forEach(tarIdxName -> {
+            if (!srcIdxNames.contains(tarIdxName)) {
+                // index to be dropped
+                DBObjectComparisonResult result = new DBObjectComparisonResult(DBObjectType.INDEX, tarIdxName,
+                        this.sourceSchemaName, this.targetSchemaName);
+                result.setComparisonResult(ComparisonResult.ONLY_IN_TARGET);
+                result.setChangeScript(
+                        appendDelimiterIfNotExist(this.targetTableEditor.getIndexEditor()
+                                .generateDropObjectDDL(tarIdxMapping.get(tarIdxName))));
+                returnVal.add(result);
+            }
+        });
+
+        srcIdxNames.forEach(srcIdxName -> {
+            DBObjectComparisonResult result = new DBObjectComparisonResult(DBObjectType.INDEX, srcIdxName,
+                    this.sourceSchemaName, this.targetSchemaName);
+            DBTableIndex copiedSrcIdx = new DBTableIndex();
+            BeanUtils.copyProperties(srcIdxMapping.get(srcIdxName), copiedSrcIdx);
+            copiedSrcIdx.setSchemaName(this.targetSchemaName);
+            if (tarIdxNames.contains(srcIdxName)) {
+                String ddl = this.targetTableEditor.getIndexEditor().generateUpdateObjectDDL(
+                        tarIdxMapping.get(srcIdxName), copiedSrcIdx);
+                if (!ddl.isEmpty()) {
+                    // index to be updated
+                    result.setComparisonResult(ComparisonResult.INCONSISTENT);
+                    result.setChangeScript(appendDelimiterIfNotExist(ddl));
+                } else {
+                    result.setComparisonResult(ComparisonResult.CONSISTENT);
+                }
+            } else {
+                // index to be created
+                result.setComparisonResult(ComparisonResult.ONLY_IN_SOURCE);
+                result.setChangeScript(appendDelimiterIfNotExist(
+                        this.targetTableEditor.getIndexEditor().generateCreateObjectDDL(copiedSrcIdx)));
+            }
+            returnVal.add(result);
+        });
+
+        return returnVal;
+    }
+
+    private List<DBObjectComparisonResult> compareTableConstraints(List<DBTableConstraint> srcTabCons,
```

Comment:
compareTablePartition
compareTableConstraints
compareTableIndexes
compareTableColumns

these method get the similar structure which is the key point that we can abstract